doc/developer/design/20230806_durable_catalog_state.md

```diff
@@ -0,0 +1,419 @@
+    /// all migrations are the same. There's probably a better way to handle this
+    /// than storing it in the catalog indefinitely.
+    fn boot_ts(&self) -> mz_repr::Timestamp;
+
+    /// Returns the epoch of the current durable catalog state. The epoch acts as
+    /// a fencing token to prevent split brain issues across two
+    /// [`DurableCatalogState`]s. When a new [`DurableCatalogState`] opens the
+    /// catalog, it will increment the epoch by one (or initialize it to some
+    /// value if there's no existing epoch) and store the value in memory. It's
+    /// guaranteed that no two [`DurableCatalogState`]s will return the same value
+    /// for their epoch.
+    ///
+    /// None is returned if the catalog hasn't been opened yet.
+    ///
+    /// NB: We may remove this in later iterations of Pv2.
+    fn epoch(&mut self) -> Option<NonZeroI64>;
+
+    // Read
+
+    /*
+     * Each object type will an accessor method of the form
+     * get_<object-type>s(&self) -> Vec<ObjectType>;
+     */
+    /// Get all clusters.
+    fn get_clusters(&mut self) -> Result<Vec<Cluster>, Error>;
+
+    /*
+     * Currently, there isn't much need for methods of the form
+     * get_<object-type>_by_id(&self, id: I) -> ObjectType;
+     * As we separate the catalog state out from the rest of the
+     * Coordinator, we will most likely need to add these methods
+     * which should be fairly straight forward.
+     */
+
+    // ...
+
+    // Write
+
+    /// Creates a new durable catalog state transaction.
+    fn transaction(&mut self) -> Transaction;
+
+    /// Commits a durable catalog state transaction.
+    fn commit_transaction(&mut self, tx: Transaction) -> Result<(), Error>;
+
+    /*
+     * The majority of writes will go through a Transaction.
+     * However as an optimization it may be useful to have more
+     * targeted methods. We can add those on a case-by-case basis.
+     */
+
+    /// Persist mapping from system objects to global IDs and fingerprints.
+    fn set_system_object_mapping(
+        &mut self,
+        mappings: Vec<SystemObjectMapping>,
+    ) -> Result<(), Error>;
+
+    // ...
+
+    // Misc
+
+    /// Confirms that this catalog is connected as the current leader.
+    ///
+    /// NB: We may remove this in later iterations of Pv2.
+    fn confirm_leadership(&mut self) -> Result<(), Error>;
+
+    /// Dumps the entire catalog contents in human readable JSON.
+    fn dump(&self) -> Result<String, Error>;
+}
+```
+
+We will also add the following methods to initialize the durable catalog state structs. They do
+not open the catalog, but they may create connections over the network.
+
+```Rust
+fn init_stash_catalog_state(
+    now: NowFn,
+    stash_factory: StashFactory,
+    stash_url: String,
+    tls: MakeTlsConnector,
+) -> Box<dyn DurableCatalogState> {
+    // ...
+}
+
+fn init_persist_catalog_state(
+    now: NowFn,
+    shard_id: ShardId,
```

Comment:
feel free to push back if this is too in the weeds, but how do we plan to pick this shard_id? every other shard_id is generated as a random uuid and written to the catalog before use but... not sure how that would work here

the obvious answer is a fixed shard_id, but it's a bit of a bummer to lose the property that shard_ids from two envs (e.g. in local dev) don't overlap

Reply:
I don't think it is too in the weeds and I have been thinking about this. Not only do we need a single unique shard ID for production, but during testing we'd like to spin up multiple concurrent environments so we'd need multiple shard IDs to spin up concurrently.

I think a good solution would be to reserve some prefix of shard IDs in persist for catalog usage. Then the coord can randomly generate shard ID suffixes. We can use the organization ID to seed the suffix generation so it's deterministic for an environment.

There is still a slight chance that two orgs in prod use the same shard ID. Will that cause issues in persist or are shard IDs isolated across environments?

Reply:
Also, a concern is how to we ensure that no shard is already using that prefix, and what do we do if a shard is using that prefix? We would maybe need a one time migration to move all shards off that prefix.

Reply:
I realized a second reason that we might want them to be different per env is that it's a nice property in persist monitoring for every shard to be different, even across envs.

I'm skeptical that reserving a prefix of shard ids is feasible. But, we're probably overthinking this. There already must be some place in crdb that an env knows to look for the catalog stash, we could just stick a shard id next to that.

Reply:
I believe the equivalent here would be the database and schema used in CRDB. @mjibson Do you know how we select and remember what database/schema an environment is using?

Reply:
My understanding is that the database and schema get passed in by cloud via the `adapter_stash_url`

https://github.com/MaterializeInc/materialize/blob/2d71ae0581cc86d03f8345402d606e49f31b1013/src/environmentd/src/bin/environmentd/main.rs#L451-L454

but I'm not really sure how they determine the URL or how they save it.

Reply:
in prod, the schema is the env name

Reply:
It looks like the database name is resolved to `environment_{kubernetes resource name}` 
https://github.com/MaterializeInc/cloud/blob/89b36e68e407fda6ca179d7607d4383f1aed51db/src/environment/src/environment_assignment.rs#L90-L92
but I don't really understand the inner workings of this in cloud.

Reply:
I asked cloud (https://materializeinc.slack.com/archives/C0246GEHL8N/p1693504735562559) the database name is deterministic: `environment_{org-id}_0`. So there's no need to write it down anywhere. 

So I don't think we could just write a shard ID down like you suggested. We could pick a shard ID via something like `hash(org-id)`, but we'd still have to deal with collisions.

Reply:
oh, `hash(org-id)` is a good idea! i'm not sure there'd be any more collisions than we'd have today?

Reply:
I added a blurb about this to the persist implementation section.